src/main/java/org/veupathdb/service/eda/ds/plugin/standalonemap/markers/QuantitativeAggregateConfiguration.java

```diff
@@ -89,4 +77,49 @@ public String serializeAverage(double d) {
       return Double.toString(d);
     }
   }
+
+  public static class CategoricalProportionAggregatorFactory implements ContinuousAggregators.ContinuousAggregatorFactory {
+    private final CategoricalAggregationConfig categoricalConfig;
+    private final Supplier<List<String>> vocabSupplier;
+    private final Set<String> distinctDenominatorValues;
+    private final Set<String> numeratorMatchSet;
+    boolean negationMode;
+
+    public CategoricalProportionAggregatorFactory(CategoricalAggregationConfig categoricalConfig,
+                                                  Supplier<List<String>> vocabSupplier) {
+      Set<String> numeratorValues = new HashSet<>(categoricalConfig.getNumeratorValues());
+      Set<String> vocabulary = new HashSet<>(vocabSupplier.get());
+      Set<String> numeratorValuesNegation = Sets.difference(vocabulary, numeratorValues);
+      // Count one of the following depending on which is cheaper:
+      // 1. Total numerator matches.
+      // 2. Everything that doesn't match the numerator.
+      if (numeratorValuesNegation.size() < numeratorValues.size()) {
+        this.negationMode = true;
+        this.numeratorMatchSet = numeratorValuesNegation;
+      } else {
+        this.negationMode = false;
+        this.numeratorMatchSet = numeratorValues;
+      }
+      Set<String> denominatorValues = new HashSet<>(categoricalConfig.getDenominatorValues());
+      this.distinctDenominatorValues = Sets.difference(denominatorValues, numeratorValues);
+      this.categoricalConfig = categoricalConfig;
+      this.vocabSupplier = vocabSupplier;
```

Comment:
Since you already have the vocab in hand, why not save that off instead of the supplier?  Not clear here if the supplier caches the values; for all we know here it could be expensive to fetch vocab (which is why there's a supplier).